### PR TITLE
[CP] Migrate MLA prefill CP (DeepSeek V3) to CP-v2 zigzag strategy

### PR DESCRIPTION
**Pull request by kpham-sgl**
## Motivation

Part of the Prefill CP refactor roadmap (#27252) and the Context Parallelism roadmap (#21788). #28421 landed the CP-v2 `ZigzagCPStrategy` for GQA/MHA (Qwen3) and the model-agnostic eager-runner boundary. This PR ports **MLA prefill CP for DeepSeek V3/R1** (originally #23292) onto that abstraction and makes the strategy own the MLA latent all-gather.

`DeepseekV3ForCausalLM` is added to `CP_V2_DEFAULT_MODEL_CLASSES`, so CP-v2 is its default (like `Qwen3MoeForCausalLM`). The legacy cp_utils v1 path stays intact and reachable via `SGLANG_ENABLE_CP_V2=0`.

## Modifications

**Boundary migration (v1↔v2 coexist):**
- `layers/cp/utils.py` — add `DeepseekV3ForCausalLM` to `CP_V2_DEFAULT_MODEL_CLASSES`.
- `models/deepseek_v2.py` / `models/deepseek_nextn.py` — gate the in-model CP entry split + exit gather to v1-only (`use_cp_v1`); the eager runner owns them under v2.
- `layers/attention/flashattention_backend.py` — MLA CP attention dispatches through `cp_strategy.run_attention` under v2, else the v1 `cp_attn_forward_extend`.
- `model_executor/runner/eager_runner.py` — CP-v2 extend is a single self-contained `_execute_extend_cp_v2` (shard at the boundary, run the body on the rank-local slice, gather before logits).

**Decision A — strategy owns the MLA latent all-gather (v2):**
- `models/.../forward_mla.py` — skip the upstream `rebuild_cp_kv_cache` under v2 (k_nope/k_pe stay rank-local).
- `flashattention_backend.py` — the MLA KV write dispatches to `cp_strategy.materialize_full_mla_kv` under v2 (gather rank-local latent → full, then `set_mla_kv_buffer`); v1 keeps the upstream-rebuilt full write.
- `layers/cp/zigzag.py` — add `materialize_full_mla_kv` (fuse compressed-KV + rope, one all-gather via `gather_kv_cache`, split back for the paged write).

**Already handled (no change needed):** the `seq_lens_k` padding flagged on #28421 — zigzag `build_metadata` bakes `pad_len` (#28421) and the FA backend widens `page_table` by `pad_delta` (#23292). Decision A's gather preserves the same padded length.

## Accuracy Tests

Existing `test/registered/cp/test_deepseek_v3_cp_single_node.py` (`tp=8, dp=2, attn-cp=4`, DeepSeek-V3-0324, GSM8k ≥ 0.935) now runs CP-v2 via the default-class enablement. Verified on 8×H200:

| Config | GSM8k |
|---|---|
| v1 (`SGLANG_ENABLE_CP_V2=0`) | 0.966 |
| v2 + Decision A (strategy latent gather) | 0.96 / 0.962 |
| v2, final (eager-runner extraction) | **0.964** |

Non-regression: v1 MLA CP unchanged (0.966); NSA/DSA and MHA CP paths are no-ops under default `SGLANG_ENABLE_CP_V2=0`.

## Out of scope (follow-ups)

- **Kimi K2.5** — CP-v2 for the multimodal wrapper (resolving its inner `language_model`) is deferred to a follow-up PR.
- MTP (nextn) + CP-v2 and DSA + CP-v2 remain out of scope (v1-only `use_cp_v1` gates make v2 a no-op there).

## Checklist

- [x] Format with pre-commit
- [x] Accuracy verified for DeepSeek V3 (see table)
- [ ] Speed/profiling (TODO)

---
### CI States

Latest PR Test (Base): :x: [Run #29629808219](https://github.com/sgl-project/sglang/actions/runs/29629808219)
Latest PR Test (Extra): :x: [Run #29629808052](https://github.com/sgl-project/sglang/actions/runs/29629808052)

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -2554,9 +2547,11 @@ def forward(
             else None
         )
 
-        if dsa_use_prefill_cp(
-            forward_batch, self.dsa_enable_prefill_cp
-        ) or mla_use_prefill_cp(forward_batch, self.mla_enable_prefill_cp):
+        if (
+            dsa_use_prefill_cp(forward_batch, self.dsa_enable_prefill_cp)
+            or mla_use_prefill_cp(forward_batch, self.mla_enable_prefill_cp)
+        ) and not is_cp_v2_active(forward_batch):
```
Comment (kpham-sgl) [inline python/sglang/srt/models/deepseek_v2.py]: grouped this into `use_cp_v1`

File: python/sglang/srt/models/deepseek_nextn.py [hunk 2]
```diff
@@ -251,9 +252,11 @@ def forward(
...
                     hidden_states = self.eh_proj(eh_input)
 
-            use_cp = dsa_use_prefill_cp(
-                forward_batch, self.dsa_enable_prefill_cp
-            ) or mla_use_prefill_cp(forward_batch, self.mla_enable_prefill_cp)
+            # CP-v2 shards/gathers at the eager-runner boundary instead.
+            use_cp = (
+                dsa_use_prefill_cp(forward_batch, self.dsa_enable_prefill_cp)
+                or mla_use_prefill_cp(forward_batch, self.mla_enable_prefill_cp)
+            ) and not is_cp_v2_active(forward_batch)
```
Comment (kpham-sgl) [inline python/sglang/srt/models/deepseek_nextn.py:259]: same below

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (kpham-sgl) [reply python/sglang/srt/models/deepseek_v2.py]: Done in ab8390fa2 — computed `use_cp_v1` once and reused it for both the entry split and the exit gather.

File: python/sglang/srt/models/deepseek_nextn.py (hunk 2 above)
Comment (kpham-sgl) [reply python/sglang/srt/models/deepseek_nextn.py:259]: Done in ab8390fa2 — renamed `use_cp` -> `use_cp_v1` here to match deepseek_v2.py.

**Comment by kpham-sgl** [comment]
I will defer Kimi changes to another PR. Need to come up with a clean way for MM models

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
MLA/DSA tests passed here
https://github.com/sgl-project/sglang/actions/runs/29629808052/job/88041219215
